walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,101 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  // comparator using the Java 8 lambda syntax
+  private final Comparator<Event> ORDER_END =
+      (Event a, Event b) -> Long.compare(a.getWhen().end(), b.getWhen().end());
+
+  public Collection<TimeRange> query(Collection<Event> events,
+                                     MeetingRequest request) {
+    /* Algorithm overview:
+    Remove people from the event list who are in the list but not in the
+    meeting. Sort all of events in descending order by end time. Grab the start
+    time for this current event and traverse through array all events that have
+    an end time greater than this start time. The time slot that can not be used
+    is then the end time of the current iteration and the minimum start time of
+    those other events. Now take the event that after the last event that was
+    discovered still within the constraints and repeat from there. Such an
+    approach will take O(nlogn) time to sort and O(n) time to iterate. The
+    removal will take O(mn) time giving us a complexity of O(mn
+    + nlogn).
+    */
+
+    // Store the times when we cannot meet, essentially take the inverse of this
+    // to determine the times we can meet.
+
+    Collection<String> meetingAttendees = request.getAttendees();
+    ArrayList<TimeRange> fullTimes = new ArrayList<>();
+    Collection<TimeRange> toReturn = new ArrayList<>();
+    // Change to ArrayList so we can sort.
+    ArrayList<Event> eventList = new ArrayList<>(events);
+    // Remove an event if none of the members in the meeting attended it.
+    eventList.removeIf(event
+                       -> !event.getAttendees().stream().anyMatch(
+                           attendee -> meetingAttendees.contains(attendee)));
+    Collections.sort(eventList, ORDER_END.reversed());
```

Comment:
It seems to me that this algorithm would work just as well in reverse (so, traversing the events chronologically). Am I missing something? In general I think algorithms are easier to understand if you stick to people's intuitive orderings, so I'd recommend working forwards through time unless there's a good reason not to.

Reply:
That's a good point, as the later part of this walkthrough builds on this part of the project and I've already made progress on that, I'll address this in a future pull request